auth/providers/azure/pop_tokenverifier_test.go

```diff
@@ -237,65 +240,140 @@ func GeneratePoPToken(ts int64, hostName, kid string) (string, error) {
 	_, err = verifier.ValidatePopToken(invalidToken)
 	assert.EqualError(t, err, "could not retrieve 'cnf' claim from access token")
 
-	invalidToken, _ = GeneratePoPToken(time.Now().Unix(), "testHostname", atCnfClaimWrong)
+	invalidToken, _ = GeneratePoPToken(time.Now().Unix(), "testHostname", atCnfClaimWrong, "")
 	_, err = verifier.ValidatePopToken(invalidToken)
 	assert.EqualError(t, err, "PoP token validate failed: 'cnf' claim mismatch")
+
+	verifierNonce := NewPoPVerifier("testHostname", 1*time.Second)
+	verifierNonce.mapCacheRetentionBuffer = 0
+	// Testing map cache cleanup
+	validToken, _ = GeneratePoPToken(time.Now().Unix(), "testHostname", "", "randomnonce1")
+	_, err = verifierNonce.ValidatePopToken(validToken)
+	assert.NoError(t, err)
+	// Wait for cache to be cleaned up
+	time.Sleep(2 * time.Second)
+
+	_, err = verifierNonce.ValidatePopToken(validToken)
+	assert.NoError(t, err)
+}
+
+// TestPopTokenVerifier_VerifySameNonce tests that the nonce is not reused using same token with different timestamp
+func TestPopTokenVerifier_VerifySameNonce(t *testing.T) {
+	nonceClaim := "nonceTest"
+	hostnameClaim := "testHostname"
+	getTokens := func() (string, string, error) {
+		popKey, err := NewSWPoPKey()
+		if err != nil {
+			return "", "", fmt.Errorf("Failed to generate Pop key. Error:%+v", err)
+		}
+
+		key, err := NewSwkKey()
+		if err != nil {
+			return "", "", fmt.Errorf("Failed to generate SF key. Error:%+v", err)
+		}
+
+		cnf := popKey.KeyID()
+		accessTokenData := fmt.Sprintf(popAccessToken, time.Now().Add(time.Minute*5).Unix(), cnf)
+		at, err := key.GenerateToken([]byte(accessTokenData))
+		if err != nil {
+			return "", "", fmt.Errorf("Error when generating token. Error:%+v", err)
+		}
+		// Hardcoding nonce to test cache
+		nonce := strings.Replace(nonceClaim, "-", "", -1)
+		keyID := popKey.KeyID()
+		algo := popKey.Alg()
+		typ := "pop"
+		header := fmt.Sprintf(`{"typ":"%s","alg":"%s","kid":"%s"}`, typ, algo, keyID)
+		headerB64 := base64.RawURLEncoding.EncodeToString([]byte(header))
+		var tokens []string
+		for i := 0; i < 2; i++ {
+			payload := fmt.Sprintf(`{ "at" : "%s", "ts" : %d, "u": "%s", "cnf":{"jwk":%s}, "nonce":"%s"}`, at, time.Now().Unix(), hostnameClaim, popKey.Jwk(), nonce)
+			payloadB64 := base64.RawURLEncoding.EncodeToString([]byte(payload))
+			h256 := sha256.Sum256([]byte(headerB64 + "." + payloadB64))
+			signature, err := popKey.Sign(h256[:])
+			if err != nil {
+				return "", "", fmt.Errorf("Error while signing pop key. Error:%+v", err)
+			}
+			signatureB64 := base64.RawURLEncoding.EncodeToString(signature)
+			tokens = append(tokens, headerB64+"."+payloadB64+"."+signatureB64)
+			// enforcing different timestamp
+			time.Sleep(1 * time.Second)
+		}
+		return tokens[0], tokens[1], nil
+	}
+	token1, token2, err := getTokens(time.Now().Unix())
+	assert.NoError(t, err)
+
+	verifier := NewPoPVerifier(hostnameClaim, 1*time.Second)
+	_, err = verifier.ValidatePopToken(token1)
+	assert.NoError(t, err)
```

Comment:
This should fail because the token is only supposed to be valid for 1 second and it's already been at least 1 second since it was generated. Try changing the token TTL in the call to `NewPoPVerifier` (but also make sure the current test fails).

Reply:
actually I think there is a bug in current implementation. Taking a look.